bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/profiles/ProfileTypeBuilder.java

```diff
@@ -0,0 +1,145 @@
+
+    private ProfileTypeBuilder(ProfileTypeUID profileTypeUID, ChannelKind channelKind) {
+        this.profileTypeUID = profileTypeUID;
+        this.channelKind = channelKind;
+    }
+
+    /**
+     * Obtain a new builder for a {@link StateProfileType} instance.
+     *
+     * @param profileTypeUID the {@link ProfileTypeUID}
+     * @return the new builder instance
+     */
+    public static ProfileTypeBuilder<StateProfileType> newState(ProfileTypeUID profileTypeUID) {
+        ProfileTypeBuilder<StateProfileType> ret = new ProfileTypeBuilder<>(profileTypeUID, ChannelKind.STATE);
+        return ret;
+    }
+
+    /**
+     * Obtain a new builder for a {@link TriggerProfileType} instance.
+     *
+     * @param profileTypeUID the {@link ProfileTypeUID}
+     * @return the new builder instance
+     */
+    public static ProfileTypeBuilder<TriggerProfileType> newTrigger(ProfileTypeUID profileTypeUID) {
+        ProfileTypeBuilder<TriggerProfileType> ret = new ProfileTypeBuilder<>(profileTypeUID, ChannelKind.TRIGGER);
+        return ret;
+    }
+
+    /**
+     * Define the human-readable label
+     *
+     * @param label
+     * @return the builder itself
+     */
+    public ProfileTypeBuilder<T> withLabel(String label) {
+        this.label = label;
+        return this;
+    }
+
+    /**
+     * Declare that the given item type(s) are supported by a profile of this type.
+     *
+     * @param itemType
+     * @return the builder itself
+     */
+    public ProfileTypeBuilder<T> withSupportedItemTypes(String... itemType) {
+        supportedItemTypes.addAll(Arrays.asList(itemType));
+        return this;
+    }
+
+    /**
+     * Declare that the given item type(s) are supported by a profile of this type.
+     *
+     * @param itemType
+     * @return the builder itself
+     */
+    public ProfileTypeBuilder<T> withSupportedItemTypes(Collection<String> itemTypes) {
+        supportedItemTypes.addAll(itemTypes);
+        return this;
+    }
+
+    /**
+     * Declare that the given channel type(s) are supported by a profile of this type.
+     *
+     * @param channelTypeUIDs
+     * @return the builder itself
+     */
+    public ProfileTypeBuilder<T> withSupportedChannelTypeUIDs(ChannelTypeUID... channelTypeUIDs) {
+        supportedChannelTypeUIDs.addAll(Arrays.asList(channelTypeUIDs));
+        return this;
+    }
+
+    /**
+     * Declare that the given channel type(s) are supported by a profile of this type.
+     *
+     * @param channelTypeUIDs
+     * @return the builder itself
+     */
+    public ProfileTypeBuilder<T> withSupportedChannelTypeUIDs(Collection<ChannelTypeUID> channelTypeUIDs) {
+        supportedChannelTypeUIDs.addAll(channelTypeUIDs);
+        return this;
+    }
+
+    /**
+     * Create a profile type instance with the previously given parameters.
+     *
+     * @return the according subtype of
+     */
+    @SuppressWarnings("unchecked")
+    public T build() {
+        final String lbl = label;
```

Comment:
Why is the label not a constructor parameter to the builder, if it apparently is mandatory for the construction of a profiletype?

Reply:
The whole intention of the builder is to not end up in a constructor-overload-mess when adding additional parameters later on. Therefore I put only the identifier there. 
That was the reason but I don't have any strong feelings about it. In the end it's a matter of taste, I guess...

Reply:
@sjka I agree that we should prevent a "constructor-overload-mess". IMHO there should be one private constructor that consumes all mandatory fields. All mandatory fields should be supplied already by the static factory methods.

Reply:
Yes, that what I meant (regardless of whether they are static factory methods or the constructor itself). Nevertheless, I count it as being overruled here. Will amend it accordingly.

Reply:
I like to understand the different approaches.

@sjka if we introduce a new mandatory field some time and do not change the "construction" (regardless if new or a static method) methods, the usage will still compile and seems to be valid. So, no error on compilation. But as soon as the build method checks the mandatory fields at runtime all unchanged (still successful compiled) usage will run into errors at runtime.

Wouldn't it better to catch such errors (they will occur) already on compile time?

To prevent big fat construction methods there are also a step builder pattern (at least I found it here "https://stackoverflow.com/a/35847487").

Reply:
I obviously didn't think it through completely - of course it would be better to fail loudly at compile time! It's an "incompatible" change, therefore implementors must notice it rather sooner than later.

You convinced me. Case closed. 😄